### ISSUE TITLE
请教一下预测代码是怎么编写和执行？

### ISSUE THREAD

**Issue by lizhangtan** [lyccol/CodeBERT-based-webshell-detection#2]
我使用自己的黑白样本数据集跑通了train.py，生成了一些cls_model_XX.pth文件，想请教一下如何使用测试数据解析预测呢？我看train.py和其他demo脚本好像都没有涉及，麻烦不吝赐教，谢谢！

**Comment by lyccol** [comment]
```
    model = NNModel.CodeBERTClassifer().cpu()
    model.load_state_dict(torch.load('model/cls_model_0.pth'))


    outputs = model(ids, mask, token_type_ids)


    print(outputs)
```

@lizhangtan 
[第61行](https://github.com/lyccol/CodeBERT-based-webshell-detection/blob/main/demo2.py)

**Comment by lizhangtan** [comment]
> ```
>     model = NNModel.CodeBERTClassifer().cpu()
>     model.load_state_dict(torch.load('model/cls_model_0.pth'))
> 
> 
>     outputs = model(ids, mask, token_type_ids)
> 
> 
>     print(outputs)
> ```
> 
> @lizhangtan [第61行](https://github.com/lyccol/CodeBERT-based-webshell-detection/blob/main/demo2.py)

谢谢提醒哈。我跑了一下demo2的代码，输入了两个用例，都是一句话的webshell，然后用训练好的模型运行了一下，最后的输出分别是这样的：
```
initing model...
open file.....
<?php eval($_POST['c'])?>
tensor([[0.1087, 0.4088]], grad_fn=<AddmmBackward0>)
```

```
initing model...
open file.....
<?$_POST['sa']($_POST['sb']);?>
tensor([[-0.0503,  0.0501]], grad_fn=<AddmmBackward0>)
```

@lyccol 想请教一下，这两个输出tensor([[0.1087, 0.4088]]；tensor([[-0.0503,  0.0501]]怎么理解或者说怎么看出模型的判定结果呢，我的理解是输出一个黑或者白的概率值。

**Comment by lyccol** [comment]
首先在[dataSet.py](https://github.com/lyccol/CodeBERT-based-webshell-detection/blob/42e5a23903253080ce9600c5ba98a306c47d69e5/dataSet.py#L64)的64行，定义黑样本为1，白样本为0.

其次再train.py中，可以看到如何取模型的输出值。
如：

```
pred = model(ids, mask, token_type_ids)
pred_choice = pred.max(1)[1]
```

针对您给的模型输出，演示如下：

```
>>> out = torch.tensor([[-0.0503,  0.0501]])
>>> out.max(1)[1]
tensor([1])
>>> out = torch.tensor([[0.1087, 0.4088]])
>>> out.max(1)[1]
tensor([1])
```

可看到模型对两个用例的选择是1，代表模型认为是黑样本。


此外，这个输出并不是概率值，只是模型输出的结果。
你可以通过torch.nn.Sigmoid()来输出概率。
具体如下：
```
>>> s = torch.nn.Sigmoid()
>>> out = torch.tensor([[0.1087, 0.4088]])
>>> out.max(1)[1]
tensor([1])
>>> s(out)
tensor([[0.5271, 0.6008]])
```
即，模型认为60.08%的概率，是黑样本.

**Comment by lizhangtan** [comment]
> 首先在[dataSet.py](https://github.com/lyccol/CodeBERT-based-webshell-detection/blob/42e5a23903253080ce9600c5ba98a306c47d69e5/dataSet.py#L64)的64行，定义黑样本为1，白样本为0.
> 
> 其次再train.py中，可以看到如何取模型的输出值。 如：
> 
> ```
> pred = model(ids, mask, token_type_ids)
> pred_choice = pred.max(1)[1]
> ```
> 
> 针对您给的模型输出，演示如下：
> 
> ```
> >>> out = torch.tensor([[-0.0503,  0.0501]])
> >>> out.max(1)[1]
> tensor([1])
> >>> out = torch.tensor([[0.1087, 0.4088]])
> >>> out.max(1)[1]
> tensor([1])
> ```
> 
> 可看到模型对两个用例的选择是1，代表模型认为是黑样本。
> 
> 此外，这个输出并不是概率值，只是模型输出的结果。 你可以通过torch.nn.Sigmoid()来输出概率。 具体如下：
> 
> ```
> >>> s = torch.nn.Sigmoid()
> >>> out = torch.tensor([[0.1087, 0.4088]])
> >>> out.max(1)[1]
> tensor([1])
> >>> s(out)
> tensor([[0.5271, 0.6008]])
> ```
> 
> 即，模型认为60.08%的概率，是黑样本.

@lyccol 非常感谢您及时和耐心的指导，按照上述说明实践之后，确实可以输出了如上所述的预测结果。后续对于模型的误报、漏报评估，可能需要基于自己的数据集进行更多的研究，再次感谢代码的分享和指导！
